rs/rust_canisters/random_traffic_test/src/main.rs

```diff
@@ -114,121 +117,184 @@ fn instructions_count_range() -> RangeInclusive<u32> {
+                call_tree_id,
+                call_depth,
+                sent_bytes: msg.count_bytes() as u32,
+                reply: None,
+            },
+        );
     });
 
+    let future = api::call::call_raw(receiver.into(), "handle_call", Encode!(&msg).unwrap(), 0);
+
+    Some((future, index))
+}
+
+/// Updates the record at `index` using the `response` to the corresponding call.
+///
+/// Removes the record for a synchronous rejection since those can be quite numerous when the
+/// subnet is at its limits. Note that since the call `index` is part of the records, removing
+/// the records for synchronous rejections will result in gaps in these numbers thus they are
+/// still included indirectly.
+fn update_record(response: &api::call::CallResult<Vec<u8>>, index: u32) {
     // Updates the `Reply` at `index` in `RECORDS`.
     let set_reply_in_call_record = move |reply: Reply| {
-        RECORDS.with_borrow_mut(|records| {
-            records[index].reply = Some(reply);
-        });
+        RECORDS.with_borrow_mut(|records| records.get_mut(&index).unwrap().reply = Some(reply));
     };
-
-    let error_code = api::call_with_callbacks(
-        receiver,
-        "handle_call",
-        &vec![0_u8; payload_bytes as usize][..],
-        move || {
-            set_reply_in_call_record(Reply::Bytes(api::arg_data().len() as u32));
-            on_response();
-        },
-        move || {
-            set_reply_in_call_record(Reply::AsynchronousRejection(
-                api::reject_code(),
-                api::reject_message(),
-            ));
-            on_response();
-        },
-    );
-    if error_code == 0 {
-        Ok(())
-    } else {
-        set_reply_in_call_record(Reply::SynchronousRejection(error_code));
-        Err(())
+    match response {
+        Err((_, msg)) if synchronous_rejection(msg) => {
+            // Remove the record for synchronous rejections.
+            SYNCHRONOUS_REJECTIONS_COUNT.set(SYNCHRONOUS_REJECTIONS_COUNT.get() + 1);
+            RECORDS.with_borrow_mut(|records| records.remove(&index).unwrap());
+        }
+        Err((reject_code, msg)) => {
+            set_reply_in_call_record(Reply::Reject(*reject_code as u32, msg.to_string()));
+        }
+        Ok(result) => {
+            set_reply_in_call_record(Reply::Bytes(result.len() as u32));
+        }
     }
 }
 
-/// Samples a weighted binomial distribution to decide whether to make a reply (true) or a
-/// downstream call (false). Defaults to `true` for bad weights (e.g. both 0).
-fn should_reply_now() -> bool {
-    RNG.with_borrow_mut(|rng| {
-        WeightedIndex::new([REPLY_WEIGHT.get(), CALL_WEIGHT.get()])
-            .map_or(true, |dist| dist.sample(rng) == 0)
-    })
+/// Generates `MAX_CALLS_PER_HEARTBEAT` calls as futures. The futures are awaited one by one such
+/// that the records can be updated without awaiting all calls at once.
```

Comment:
"One by one" sounds as if we're awaiting them sequentially (i.e. in a `for` loop). It may be clearer to leave out awaiting altogether and simply say that each record is updated as soon as the corresponding call completes or fails. I.e. describe the behavior, not the implementation.